twitter/src/main/java/com/hazelcast/jet/contrib/twitter/TwitterSources.java

```diff
@@ -0,0 +1,362 @@
+     * This method differs from the other by enabling the users to choose a different Twitter Streaming API host. e.g.
+     * {@link Constants#ENTERPRISE_STREAM_HOST}
+     * <p>
+     * Example usage:
+     * <pre>{@code
+     * List<Long> userIds = new ArrayList<Long>(
+     *                 Arrays.asList(612473L, 759251L, 1367531L, 34713362L, 51241574L, 87818409L));
+     * StreamSource<String> timestampedStreamSource =
+     *              TwitterSources.timestampedStream(
+     *                      credentials,
+     *                      Constants.STREAM_HOST,
+     *                      () -> new StatusesFilterEndpoint().followings(userIds),
+     *              );
+     * Pipeline p = Pipeline.create();
+     * StreamStage<String> srcStage = p.readFrom(timestampedStreamSource);
+     * }</pre>
+     *
+     * @param credentials      a Twitter OAuth1 credentials that consists "consumerKey",
+     *                         "consumerSecret", "token", "tokenSecret" keys.
+     *                         These hosts are defined in {@link com.twitter.hbc.core.Constants}.
+     * @param host             a Twitter host URL to connect.
+     * @param endpointSupplier Supplier that supplies a Twitter StreamingEndpoint to connect to source.
+     * @return a timestamped stream source to use in {@link com.hazelcast.jet.pipeline.Pipeline#readFrom}
+     */
+
+    @Nonnull
+    public static StreamSource<String> timestampedStream(
+            @Nonnull Properties credentials,
+            @Nonnull String host,
+            @Nonnull SupplierEx<? extends StreamingEndpoint> endpointSupplier
+    ) {
+        return SourceBuilder.timestampedStream("twitter-timestamped-stream-source",
+                ctx -> new TwitterStreamSourceContext(credentials, host, endpointSupplier))
+                .fillBufferFn(TwitterStreamSourceContext::fillTimestampedBuffer)
+                .destroyFn(TwitterStreamSourceContext::close)
+                .build();
+    }
+
+    /**
+     * Creates a {@link BatchSource} which emits tweets as the form of {@link Status} by getting from
+     * Twitter's Search API for data ingestion to Jet pipelines.
+     * <p>
+     * Example usage:
+     * <pre>{@code
+     * Properties credentials = loadTwitterCredentials();
+     * BatchSource<Status> twitterSearchSource =
+     *              TwitterSources.search(
+     *                      credentials,
+     *                      "Jet flies"
+     *              );
+     * Pipeline p = Pipeline.create();
+     * BatchStage<Status> srcStage = p.readFrom(streamSource);
+     * }</pre>
+     *
+     * @param credentials a Twitter OAuth1 credentials that consists "consumerKey",
+     *                    "consumerSecret", "token", "tokenSecret" keys.
+     * @param query       a search query
+     * @return a batch source to use in {@link com.hazelcast.jet.pipeline.Pipeline#readFrom}
+     */
+    @Nonnull
+    public static BatchSource<Status> search(
```

Comment:
Could you please explain why this is a batch source ? Is there an upper limit on the results returned by Twitter on this API ?

Reply:
Twitter restricts the repeated(continuous) access to its search endpoint so one can only make 180 calls every 15 mins. Currently, the Twitter connector tries to get the search results from the search endpoint until it gets exhausted. If you think it's more appropriate,
 I can also add a limit parameter to it.

Reply:
We might add another endpoint which waits until more search permits are available when user hits the limits but leave it for future improvements.

Reply:
Also, despite the streaming nature of the tweets, the tweet searching task is inherently bounded with tweets posted so far. When a user searches for tweets, future tweets are not his/her concern.

The reason for iteratively getting search results with the same query string(performed in fillBuffer function) is the paging structure of the search API. After making a search request, the API returns only the most recent tweets with a specific window size for this query and also a cursor(max_id) that denotes (id of the oldest tweet -1) in the current batch. To get the previous batches of tweets, the same query is iteratively performed with the last returned cursor. [for more info](https://developer.twitter.com/en/docs/tweets/timelines/guides/working-with-timelines)